### ISSUE TITLE
[css-fonts-5] Make `unicode-range` syntax suck less

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#7921]
Right now `unicode-range` accepts everything in terms of codepoints. For example:

```css
/* yen, kanji, hiragana, katakana */
unicode-range: U+A5, U+4E00-9FFF, U+30??, U+FF00-FF9F;
```

This has several problems:
- It's hard to read, even if you want to specify specific characters, you need to find their codepoints
- Broader character classes (e.g. Japanese letter, Emoji, Digit) need to be specified manually, which is error-prone
- There is no exclusion syntax (all characters in the font minus these), the range needs to be tediously constructed by starting from `U+0000` and ending at `U+FFFF` breaking as needed in between.

**Comment by LeaVerou** [comment]
Ideas and related work:

> 1. It's hard to read, even if you want to specify specific characters, you need to find their codepoints

For this, I'd propose allowing `<string>` as well, and treating it as specifying the set that is derived from the union of all codepoints in the string. E.g. this would be valid:

```css
unicode-range: "&";
```

> 2. Broader character classes (e.g. Japanese letter, Emoji, Digit) need to be specified manually, which is error-prone

For this there has already been discussion in https://github.com/w3c/csswg-drafts/issues/4573 and just needs spec edits.

> 3. There is no exclusion syntax (all characters in the font minus these), the range needs to be tediously constructed by starting from `U+0000` and ending at `U+FFFF` breaking as needed in between.

Perhaps we need some kind of *not* operator. This can even be just a keyword (`not`, `exclude`?) in front of existing value syntax:

```
<urange> = <urange> | not <urange>
```

Which would allow things like:

```css
unicode-range: greek, not japanese, not U+A5;
```

Or a minus operator and a keyword for all characters? 

```css
unicode-range: greek except "π";
```

**Comment by tabatkins** [comment]
Yeah, I think `[ not? [ <urange> | string | <script-keyword>] ]#` is pretty reasonable, with strings being equivalent to a range that covers all the codepoints of the string. All positive ranges would be added, then all the negative ranges would be subtracted; I don't think there's a real need to subtract from a particular range.

**Comment by dbaron** [comment]
[From](https://github.com/w3c/csswg-drafts/issues/7921#issuecomment-1284681117) @tabatkins :
> All positive ranges would be added, then all the negative ranges would be subtracted

Presumably if there are no positive ranges, then the starting point would be all characters rather than none.

[From](https://github.com/w3c/csswg-drafts/issues/7921#issuecomment-1284346064) @LeaVerou :
> ```css
> unicode-range: greek, not japanese, not U+A5;
> ```

Were you expecting that `unicode-range: greek, not japanese` would do something different from just `unicode-range: greek`?  If so, what?  (I'm assuming that the `greek` and `japanese` character ranges don't intersect... though maybe that's an incorrect assumption.)

**Comment by LeaVerou** [comment]
> Were you expecting that `unicode-range: greek, not japanese` would do something different from just `unicode-range: greek`? If so, what? (I'm assuming that the `greek` and `japanese` character ranges don't intersect... though maybe that's an incorrect assumption.)

I was just trying to show syntax, but I agree that is a poor example. They definitely don't intersect! The fact that I can't easily think of examples that _do_ intersect probably proves that `not` should just apply to the whole value. We should however design the syntax so that it can be extended in the future if such use cases emerge. The syntax @tabatkins proposes cannot be extended in the future, and is confusing to read due to the precedence of commas vs words in natural language (and most places in CSS as well):

```css
unicode-range: not greek, japanese, U+A5;
```

It looks like `not` applies to `greek`, but it actually applies to the whole value. A functional syntax can fix this and make the syntax both clearer and more extensible:

```css
unicode-range: not(greek, japanese, U+A5);
```

**Comment by tabatkins** [comment]
> It looks like not applies to greek, but it actually applies to the whole value.

I think you misread my grammar, that's absolutely not correct. The multiplier is on the outside, covering the `not? <range-options>` part. `not greek, japanese, U+A5` definitely does mean "the japanese range, plus U+00a5, minus the greek range".

**Comment by LeaVerou** [comment]
> > It looks like not applies to greek, but it actually applies to the whole value.
> 
> I think you misread my grammar, that's absolutely not correct. The multiplier is on the outside, covering the `not? <range-options>` part. `not greek, japanese, U+A5` definitely does mean "the japanese range, plus U+00a5, minus the greek range".

Oh you're right! I edited my comment.

**Comment by aphillips** [comment]
You might want to start by looking at what Unicode and ICU have done in this space. For example, the [UnicodeSet](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/text/UnicodeSet.html) class in ICU4J is similar to the kinds of "range selection" you're describing here--one can add characters according to various Unicode properties, classes, and scripts to build up ranges, invert ranges, etc.

I think the descriptions in the thread above need to be tighter. Are `greek` and `japanese` supposed to be script names, e.g. equivalent to ISO15924 codes like `Grek` and `Jpan`? Or are they meant to describe specific character sets, such as the `el` (Greek) and `ja` locale exemplary sets in CLDR (such as [this one](https://unicode-org.github.io/cldr-staging/charts/latest/summary/el.html#2703e9d07ab2ef3a))? These kinds of sets definitely *do* intersect in various ways (and most languages use at least some of the "common" script--think punctuation). I'll also call out that Unicode runs all the way to `U+10FFFF`.

**Comment by r12a** [comment]
> For this, I'd propose allowing <string> as well, and treating it as specifying the set that is derived from the union of all codepoints in the string. E.g. this would be valid:
> 
> unicode-range: "&";

It may also be useful (i haven't thought it through completely) to allow ranges separated by hyphens, like:

`unicode-range: "&¡-§©"`

which would include the characters &¡¢£¤¥¦§©.  You'd need a way of escaping the - character though.

(Two other cautions about situations where it may be better to stick with code point numbers: 

[1] Using characters instead of code point values may cause some difficulty when specifying RTL character sets.  For example in

unicode-range: "ذ-خ", "ى", "a-z", "ب-ت";

the underlying order is not what you see (although it could be worse).

[2] You'll probably still want to use code point values for combining characters and invisible characters, and especially for formatting characters such as RLI/LRI etc which will again make the declaration look odd.
)

**Comment by tabatkins** [comment]
Right, those issues are precisely why I *don't* think we want to allow string-based ranges, at least not with that syntax. A range(start, end) function could potentially work, if needed. (Tho since all the syntactic characters inside the parens are non-directional it still ends up being very confusingly visibly reordered if viewed in a web-based editor.)

**Comment by LeaVerou** [comment]
I think ranges are useful but obviously the token that indicates this is a range would need to be outside the string. Eg a function like @tabatkins described or even `<string> to <string> `

**Comment by r12a** [comment]
> the token that indicates this is a range would need to be outside the string

Not necessarily.  On the (probably rare) occasion where - has to be specified as a character it could be escaped (like in regex expressions).  In fact, this whole thing sounds very like establishing a regex expression, so perhaps that offers an alternative approach to the syntax?  

That would also allow mixing of characters and code point values, eg. if a range you specify starts with a visible character but ends with an invisible one.

**Comment by LeaVerou** [comment]
I think that makes it harder to read what the range is. I love regex, but it's not exactly known for its readability 😀

**Comment by LeaVerou** [comment]
> That would also allow mixing of characters and code point values, eg. if a range you specify starts with a visible character but ends with an invisible one.

Not sure I follow. If anything it seems to me that doing ranges with syntax outside the string makes this easier.

**Comment by svgeesus** [comment]
@aphillips 

> I also just has a look at the text mentioned in #4573 located [here](https://drafts.csswg.org/css-fonts/#unicode-range-desc). Have we (I18N) reviewed this yet?? I think (if I had been the reviewer) I would have proposed issues against the variable width `U+` syntax and some of the wording...

That descriptor (and most of the spec text) is [from CSS2, in 1998](https://www.w3.org/TR/1998/REC-CSS2-19980512/fonts.html#dataqual) by the way.

**Comment by tabatkins** [comment]
> or even `<string> to <string>`

Playing with it a bit myself, unfortunately I think we'd be well-served by using a separator token with strong LTR directionality like `to`.

If you're trying to denote a range from U+062E (<bdi>خ</bdi>) to U+0630 (<bdi>ذ</bdi>), you get the following results with a weak directionality vs strong directionality separator:

range("خ" to "ذ")

range("خ", "ذ")

The above two strings are *exactly identical* save for the separator used, but the bidi algorithm makes the second look like it's in the wrong order.

**Comment by svgeesus** [comment]
My [abject apologies, once again](https://www.w3.org/Talks/2017/font-level-up/#apology) for the unicode-range syntax.

![image](https://github.com/w3c/csswg-drafts/assets/2506926/1a1b1296-a582-422b-86d0-e341c0ede106)

_"Put it in for now, Chris, until we come up with something better"_ -- Håkon Wium Lie, spring 1997

On the other hand, at least it _wasn't the **worst** syntax proposed._ Feast your eyes on the hex-encoded BMP bitmap:

> unicode-range: 0x02037FBC4571000003100C000000100010000300BDF74300000000000

**Comment by aphillips** [comment]
> That descriptor (and most of the spec text) is [from CSS2, in 1998](https://www.w3.org/TR/1998/REC-CSS2-19980512/fonts.html#dataqual) by the way.

I'll remind Martin and Misha that they missed one. 😆

**Comment by brianjlacy** [comment]
> _"Put it in for now, Chris, until we come up with something better"_ -- Håkon Wium Lie, spring 1997

Well hey, turns out it's had a pretty good run, now let's all just come up with something better! 😁

**Comment by svgeesus** [comment]
Related:
 - https://github.com/w3c/csswg-drafts/issues/10651
 - https://github.com/w3c/csswg-drafts/issues/4573
